storage: properly invoke endCmds.done() when abandonment succeeds

petermattis opened the pull request:
Fixes #11986

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12000)
<!-- Reviewable:end -->

petermattis committed:
storage: properly invoke endCmds.done() when abandonment succeeds

Fixes #11986

petermattis commented:
This clearly need a test.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replica.go, line 1454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KY9yLRi2yZFXpCfosLt:-KY9yLRi-7TprCseZBLu:b-rsrr1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e017068154dd363b5cc3d65f80209573c51c4eca/pkg/storage/replica.go#L1454)):*
> ```Go
> 								"waited 1m for dependencies: cmd-global:\n%s\nglobal:\n%s"+
> 									"cmd-local:\n%s\nlocal:%s\n",
> 								cmdGlobal, g, cmdLocal, l)
> ```

Not sure if we should keep this or not.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KY9yH2V5JgdzVLFEOm4:b3m39ee)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Test added. Fails 100% of the time before this PR.

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KY9yLRi2yZFXpCfosLt:-KYAN1gk30h1KsSdVOcG:biawlvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e017068154dd363b5cc3d65f80209573c51c4eca/pkg/storage/replica.go#L1454)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Not sure if we should keep this or not.
</blockquote></details>

Doesn't seem like a bad idea.  This should only happen after a chaos style event.  If we see these in the log without we should investigate.

---

*[pkg/storage/replica.go, line 1441 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAMYrM3xXLV-atttCI:-KYAMYrM3j8M2BgZe1Uo:bdnndud) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1441)):*
> ```Go
> 					// out before their prerequisites, so we still have to wait it
> 					// out.
> 					timer := time.After(time.Minute)
> ```

Can you foresee this value ever needing to be adjusted?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KYANhsd5ARIV_k4uqTO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAow6Ft5GdyiSPxnrM:-KYAow6GPsmvvQBM4kUa:bkinswo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/command_queue.go#L104)):*
> ```Go
> }
> 
> func (c *cmd) String() string {
> ```

could you add a simple test for this? would be nice to see what it looks like.

ditto for the command queue.

---

*[pkg/storage/command_queue.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYApPTweB6ugw52f2e0:-KYApPTweB6ugw52f2e1:b-yliwlk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/command_queue.go#L105)):*
> ```Go
> 
> func (c *cmd) String() string {
> 	if c == nil {
> ```

does this actually happen?

---

*[pkg/storage/replica.go, line 1436 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoKeMVDXj6Bj7eMx0:-KYAoKeMVDXj6Bj7eMx1:b-5subqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1436)):*
> ```Go
> 				log.Warning(ctx, errStr)
> 				log.ErrEvent(ctx, errStr)
> 				go func() {
> ```

is there any reason this doesn't run in a task? that would give you a child context to use inside the task rather than having to `context.Background` below.

---

*[pkg/storage/replica.go, line 1445 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAohNLp023Shj9Y4xl:-KYAohNLp023Shj9Y4xm:b-ybqzj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1445)):*
> ```Go
> 						select {
> 						case <-ch:
> 						case <-timer:
> ```

are you sure you don't want to use a ticker here? seems like you want this to print every minute rather than just once.

---

*[pkg/storage/replica.go, line 1452 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoW8rPMMOjAR_w2fG:-KYAoW8rPMMOjAR_w2fH:b-augtxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1452)):*
> ```Go
> 							r.cmdQMu.Unlock()
> 							log.Infof(r.AnnotateCtx(context.Background()),
> 								"waited 1m for dependencies: cmd-global:\n%s\nglobal:\n%s"+
> ```

nit extract "1 minute" so this doesn't rot if the value changes.

---

*[pkg/storage/replica_test.go, line 5901 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYApeQLCZ4bIXuOXtUk:-KYApeQLCZ4bIXuOXtUl:b-b1j7cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica_test.go#L5901)):*
> ```Go
> 
> 	_, pErr := tc.repl.addWriteCmd(ctx, ba)
> 	if pErr == nil {
> ```

can you check for a more specific error?

---

*[pkg/storage/replica_test.go, line 5910 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAq9KYhHj5fmagboCQ:-KYAq9KYhHj5fmagboCR:bhxjcje) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica_test.go#L5910)):*
> ```Go
> 	}
> 	tc.repl.cmdQMu.Lock()
> 	if s := tc.repl.cmdQMu.global.String(); s != "" {
> ```

can't you use `.treeSize()` instead of relying on the string representation?

---

*[pkg/storage/replica_test.go, line 5911 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYApsoXmmfDh-LN1GDp:-KYApsoXmmfDh-LN1GDq:b-h9wf9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica_test.go#L5911)):*
> ```Go
> 	tc.repl.cmdQMu.Lock()
> 	if s := tc.repl.cmdQMu.global.String(); s != "" {
> 		t.Fatalf("expected empty command queue, but found\n%s", s)
> ```

if this fails the lock will remain locked, and the test will deadlock.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KYAqOptjW1nCdgg-E9Z:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAow6Ft5GdyiSPxnrM:-KYB_RM7ohJK0VwvoRre:boclx5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/command_queue.go#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

could you add a simple test for this? would be nice to see what it looks like.

ditto for the command queue.
</blockquote></details>

That seems quite onerous for debugging code.

---

*[pkg/storage/command_queue.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYApPTweB6ugw52f2e0:-KYB_G6LA0363OgxsY5a:bex1npo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/command_queue.go#L105)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

does this actually happen?
</blockquote></details>

Not now, but I was logging a `nil` `*cmd` during debugging as that is what is returned by `CommandQueue.add` if there is nothing to block on. 

---

*[pkg/storage/replica.go, line 1454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KY9yLRi2yZFXpCfosLt:-KYB_zk-bPDOM4bXLc4g:b-x616dz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e017068154dd363b5cc3d65f80209573c51c4eca/pkg/storage/replica.go#L1454)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Doesn't seem like a bad idea.  This should only happen after a chaos style event.  If we see these in the log without we should investigate.
</blockquote></details>

I'm pretty sure this was happening on other clusters without chaos involved. When I checked earlier it seems to be happening on `delta`. I'm not sure a log message is sufficient to catch a similar problem in the future. All too often we ignore the logs. We should probably add a metric for operations that are blocked longer than a certain length of time. I'll file an issue.

---

*[pkg/storage/replica.go, line 1436 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoKeMVDXj6Bj7eMx0:-KYB__Dgzqcc0tc-_oY0:b-4yx0kv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1436)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is there any reason this doesn't run in a task? that would give you a child context to use inside the task rather than having to `context.Background` below.
</blockquote></details>

Didn't you write this code?

---

*[pkg/storage/replica.go, line 1441 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAMYrM3xXLV-atttCI:-KYB_bGDV_3EtVxswkCs:b2nhky0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1441)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you foresee this value ever needing to be adjusted?
</blockquote></details>

No.

---

*[pkg/storage/replica.go, line 1445 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAohNLp023Shj9Y4xl:-KYB_eL6K0jUgmBDwbQi:bxreinw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1445)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

are you sure you don't want to use a ticker here? seems like you want this to print every minute rather than just once.
</blockquote></details>

Printing just once was sufficient to track down the problem.

---

*[pkg/storage/replica.go, line 1452 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoW8rPMMOjAR_w2fG:-KYB_xtDC5yNTa8COazi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1452)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit extract "1 minute" so this doesn't rot if the value changes.
</blockquote></details>

Done.

---

*[pkg/storage/replica_test.go, line 5901 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYApeQLCZ4bIXuOXtUk:-KYBajSzAV5bRt6JhMXP:b-spl95r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica_test.go#L5901)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you check for a more specific error?
</blockquote></details>

You mean like a few lines down?

---

*[pkg/storage/replica_test.go, line 5910 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAq9KYhHj5fmagboCQ:-KYBapen1s7jj443DYnY:brx0y6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica_test.go#L5910)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can't you use `.treeSize()` instead of relying on the string representation?
</blockquote></details>

Does it matter? The string gives us a nice output on failure.

---

*[pkg/storage/replica_test.go, line 5911 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYApsoXmmfDh-LN1GDp:-KYBauZMGw8qsEHIXAuP:b-h4lklr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica_test.go#L5911)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

if this fails the lock will remain locked, and the test will deadlock.
</blockquote></details>

No, it doesn't deadlock or block the test. The fail is clean. I've switch to a `defer`, though it doesn't matter here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/command_queue.go, line 104 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAow6Ft5GdyiSPxnrM:-KYBdsnwYi8nv8Z-LZXd:b40wjn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/command_queue.go#L104)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

That seems quite onerous for debugging code.
</blockquote></details>

I was imagining a pretty basic, single test, but OK.

---

*[pkg/storage/replica.go, line 1436 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoKeMVDXj6Bj7eMx0:-KYBcQbxYDoes5-vueuX:bknliuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1436)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Didn't you write this code?
</blockquote></details>

I did, but:
- I didn't need a logging context
- RunAsyncTask didn't provide a context at that time

---

*[pkg/storage/replica.go, line 1445 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAohNLp023Shj9Y4xl:-KYBdmr5WJCfImtavmfL:b-8mgqor) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1445)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Printing just once was sufficient to track down the problem.
</blockquote></details>

...OK

---

*[pkg/storage/replica_test.go, line 5901 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYApeQLCZ4bIXuOXtUk:-KYBdVBRarp5XELIxnTL:b5c5laz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica_test.go#L5901)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You mean like a few lines down?
</blockquote></details>

Oh, for some reason I thought that was a different error, but right you are.

---

*[pkg/storage/replica_test.go, line 5910 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAq9KYhHj5fmagboCQ:-KYBdgB30tls5Wt-jGK9:b2nhktp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica_test.go#L5910)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Does it matter? The string gives us a nice output on failure.
</blockquote></details>

OK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1436 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoKeMVDXj6Bj7eMx0:-KYBhRhIB-v8LBuDXTEr:b-v640xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1436)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I did, but:
- I didn't need a logging context
- RunAsyncTask didn't provide a context at that time
</blockquote></details>

Good points and yet the code as written has been exercised in production. Since we don't have a test which exercises this code path any change could have a lurking bug. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1436 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoKeMVDXj6Bj7eMx0:-KYBj0-3q43V7pDdSNiL:b-t7wvme) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1436)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Good points and yet the code as written has been exercised in production. Since we don't have a test which exercises this code path any change could have a lurking bug. 
</blockquote></details>

I suppose I'll settle for a TODO. Oh, and also this code didn't originally `select` on the stopper, so a task would have blocked indefinitely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1436 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoKeMVDXj6Bj7eMx0:-KYBjQO17gsSq1gbslOe:b-psd3b7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1436)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I suppose I'll settle for a TODO. Oh, and also this code didn't originally `select` on the stopper, so a task would have blocked indefinitely.
</blockquote></details>

http://github.com/cockroachdb/cockroach/commit/ae081404307493a8951f4119f9aa945d312602f8

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/replica.go, line 1436 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYAoKeMVDXj6Bj7eMx0:-KYBvBokBCoPZ1sY3Rzy:b-fqj5vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c1ede0e1b1014f06024e32ba255948ad37efdc9/pkg/storage/replica.go#L1436)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

http://github.com/cockroachdb/cockroach/commit/ae081404307493a8951f4119f9aa945d312602f8
</blockquote></details>

Added a `TODO(tamird)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KYBy7EnrawrhkRPEc48:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but @spencerkimball should probably weigh in too.

---

Reviewed 1 of 2 files at r1, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KY9yLRi2yZFXpCfosLt:-KYCCnfkXfuOd63OGWe_:b-nzefmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e017068154dd363b5cc3d65f80209573c51c4eca/pkg/storage/replica.go#L1454)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm pretty sure this was happening on other clusters without chaos involved. When I checked earlier it seems to be happening on `delta`. I'm not sure a log message is sufficient to catch a similar problem in the future. All too often we ignore the logs. We should probably add a metric for operations that are blocked longer than a certain length of time. I'll file an issue.
</blockquote></details>

Seems like it’s worth keeping until we have these issues under control (and even then we may want to keep the code around so we can turn it on with a flag/env var). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KYCeZhDP74Q1m4zsUlc:buqnarb)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1954 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYCifpjPk_tm7vSx-Q-:-KYCifpkKzvYE9EV5CHU:b71m9e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d309b4411007fcace6cb619c7cca0f68a339f49a/pkg/storage/replica.go#L1954)):*
> ```Go
> 				log.Warningf(ctx, "context cancellation after %0.1fs of attempting command %s",
> 					timeutil.Since(startTime).Seconds(), ba)
> 				return nil, roachpb.NewError(roachpb.NewAmbiguousResultError(ctx.Err().Error())), proposalNoRetry
> ```

On second thought, I don’t think this is correct. If the command has been proposed, we can’t invoke `endCmds.done` until the command has either executed or we can be sure it won’t be. `tryAbandon` doesn’t give us that guarantee (hence the AmbiguousResultError here). I’m not sure if `tryAbandon` makes sense any more, unless we can make it more conservative so it only returns true if the command is guaranteed to never execute (in which case we can no longer “expect it to finish soon” and it wouldn’t be appropriate to continue in this method. We’d have to spin off a new goroutine like the one in `beginCmds` to wait for the command to complete or maybe to find a point when it can be definitively canceled.

---

*[pkg/storage/replica_test.go, line 5896 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYCjUgFX9kLy5Bzih3k:-KYCjUgFX9kLy5Bzih3l:b-vfl5wo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d309b4411007fcace6cb619c7cca0f68a339f49a/pkg/storage/replica_test.go#L5896)):*
> ```Go
> 	tc.repl.mu.submitProposalFn = func(*EvalResult) error {
> 		cancel()
> 		return nil
> ```

Swallowing the proposal here makes the test deterministic but means that it won’t experience certain real-world concurrency issues between the proposal and the cancelation. Proceeding to submit the proposal would I think highlight the problems I indicated above (and if it doesn’t we probably need more sanity checks in the command queue).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KYCjW5J3wKjUSkyfw4J:b8b79qu)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -1936,10 +1951,10 @@ func (r *Replica) tryAddWriteCmd(
-				endCmds = nil
 				return nil, roachpb.NewError(roachpb.NewAmbiguousResultError(ctx.Err().Error())), proposalNoRetry
 			}
+			// Set endCmds to nil because they will be invoked in processRaftCommand.
```
The code here was correct before. We cannot end the commands by exiting this method early with `endCmds` still non-nil. This would incorrectly unblock the command queue, possibly allowing overlapping commands to execute concurrently.

What needs to be changed is code in `tryAbandon`. The whole point of this `tryAbandon` mechanism is to avoid use of a context which might have already had its trace span finalized when processing the raft function. I think `tryAbandon` needs to be modified to ensure that we continue to keep the `EvalResult` in the `Replica.mu.replicas` map (with `endCmds` left set), but just change something to indicate that the context can no longer be used when `processRaftCommand` is executed.

petermattis replied:
Well, the previous code was correct in one area but broken in another. And this change is correct in one area but broken in another. I've got my head wrapped around the problem. Let me go back to the drawing board.

petermattis commented:
This is ready for another look.

---

Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KY9yLRi2yZFXpCfosLt:-KYElNMf41pd2hMfq9E7:bjebi22) ([raw file](https://github.com/cockroachdb/cockroach/blob/e017068154dd363b5cc3d65f80209573c51c4eca/pkg/storage/replica.go#L1454)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Seems like it’s worth keeping until we have these issues under control (and even then we may want to keep the code around so we can turn it on with a flag/env var). 
</blockquote></details>

Ok, I've left it in. Perhaps we can remove once #12001 is fixed.

---

*[pkg/storage/replica.go, line 1954 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYCifpjPk_tm7vSx-Q-:-KYElT5CpQ8WTP16ARFj:b1jf76g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d309b4411007fcace6cb619c7cca0f68a339f49a/pkg/storage/replica.go#L1954)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

On second thought, I don’t think this is correct. If the command has been proposed, we can’t invoke `endCmds.done` until the command has either executed or we can be sure it won’t be. `tryAbandon` doesn’t give us that guarantee (hence the AmbiguousResultError here). I’m not sure if `tryAbandon` makes sense any more, unless we can make it more conservative so it only returns true if the command is guaranteed to never execute (in which case we can no longer “expect it to finish soon” and it wouldn’t be appropriate to continue in this method. We’d have to spin off a new goroutine like the one in `beginCmds` to wait for the command to complete or maybe to find a point when it can be definitively canceled.
</blockquote></details>

Ok, I've reworked this so that `tryAbandon` only clears `cmd.Local.ctx` and adjusted `processRaftCommand` to handle that situation. 

---

*[pkg/storage/replica.go, line 1956 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KY9yDFl4fdQufBUFAX6-r4-1956:-KYElahNdI2Mr8y91m83:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d309b4411007fcace6cb619c7cca0f68a339f49a/pkg/storage/replica.go#L1956)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Well, the previous code was correct in one area but broken in another. And this change is correct in one area but broken in another. I've got my head wrapped around the problem. Let me go back to the drawing board.
</blockquote></details>

Done.

---

*[pkg/storage/replica_test.go, line 5896 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYCjUgFX9kLy5Bzih3k:-KYEleGJT1fq8h73-R_f:be3r7sa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d309b4411007fcace6cb619c7cca0f68a339f49a/pkg/storage/replica_test.go#L5896)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Swallowing the proposal here makes the test deterministic but means that it won’t experience certain real-world concurrency issues between the proposal and the cancelation. Proceeding to submit the proposal would I think highlight the problems I indicated above (and if it doesn’t we probably need more sanity checks in the command queue).
</blockquote></details>

I no longer swallow the proposal, but go through a little dance to ensure that the proposal is sent to Raft until after the cancellation is processed. The test still deterministically fails before this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KYEloDojEzqJUhtOyvT:b-2hls1n)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1977 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEmQsN-T13mn-QMyhZ:-KYEmQsN-T13mn-QMyh_:b-yc0mx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c35c44500c29e054dfd53317ea1b731925cd0c4/pkg/storage/replica.go#L1977)):*
> ```Go
> 			// TODO(peter): This is strange. It would make more sense to set `endCmds
> 			// = nil` immediately after proposing to Raft. Does this open a hole in
> 			// correctness during graceful shutdown?
> ```

Looks like not setting `endCmds = nil` here is causing various tests to fail due to calling `endCmds.done` twice. Fun.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r5.
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1977 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEmQsN-T13mn-QMyhZ:-KYEnizbuYPo8ZxptXDx:b4pc0l4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c35c44500c29e054dfd53317ea1b731925cd0c4/pkg/storage/replica.go#L1977)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Looks like not setting `endCmds = nil` here is causing various tests to fail due to calling `endCmds.done` twice. Fun.
</blockquote></details>

I'm pretty weary of leaving this here without getting to the bottom of it.

---

*[pkg/storage/replica_test.go, line 5879 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEokK42xaCqmAiV2wL:-KYEokK42xaCqmAiV2wM:b-nn28xg) ([raw file](https://github.com/cockroachdb/cockroach/blob/77e93615fe4b07015fff1fa0f7d17cc488c023c9/pkg/storage/replica_test.go#L5879)):*
> ```Go
> 	defer stopper.Stop()
> 	tc := testContext{}
> 	tc.StartWithStoreConfig(t, stopper, TestStoreConfig(nil))
> ```

`tc.Start(t, stopper)`? or do you intentionally need a non-manual clock?

---

*[pkg/storage/replica_test.go, line 5929 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEozDLIx8_ot53RIQE:-KYEozDLIx8_ot53RIQF:b1u5v5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/77e93615fe4b07015fff1fa0f7d17cc488c023c9/pkg/storage/replica_test.go#L5929)):*
> ```Go
> 		defer tc.repl.cmdQMu.Unlock()
> 		if s := tc.repl.cmdQMu.global.String(); s == "" {
> 			t.Fatal("expected non-empty command queue")
> ```

log the string if it isn't what you expect

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1977 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEmQsN-T13mn-QMyhZ:-KYEpPFkBZvv3meqGBmc:b-t2dqas) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c35c44500c29e054dfd53317ea1b731925cd0c4/pkg/storage/replica.go#L1977)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I'm pretty weary of leaving this here without getting to the bottom of it.
</blockquote></details>

Spencer claims this was necessary to avoid flaky tests. I'm not seeing that, but am trying to reproduce.

---

*[pkg/storage/replica_test.go, line 5879 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEokK42xaCqmAiV2wL:-KYEq9nh5_dySSdfjlP-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77e93615fe4b07015fff1fa0f7d17cc488c023c9/pkg/storage/replica_test.go#L5879)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`tc.Start(t, stopper)`? or do you intentionally need a non-manual clock?
</blockquote></details>

Done.

---

*[pkg/storage/replica_test.go, line 5929 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEozDLIx8_ot53RIQE:-KYEq53I4sOypEloOPLs:bmxqls6) ([raw file](https://github.com/cockroachdb/cockroach/blob/77e93615fe4b07015fff1fa0f7d17cc488c023c9/pkg/storage/replica_test.go#L5929)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

log the string if it isn't what you expect
</blockquote></details>

The string is empty. What is there to log?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_test.go, line 5929 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEozDLIx8_ot53RIQE:-KYEr32BVCzDALQzqsuo:b-4pk0ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/77e93615fe4b07015fff1fa0f7d17cc488c023c9/pkg/storage/replica_test.go#L5929)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The string is empty. What is there to log?
</blockquote></details>

derp. carry on. (but cough cough maybe use `treeSize`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I had to tweak this one more time. Setting `cmd.Local.ctx = nil` caused problems for reproposals. `tryAbandon` now creates a new context.

---

Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1957 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEs-FZymKDqvEKJC3X:-KYEs-FZymKDqvEKJC3Y:b-sltkoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfc11855544a673d2b3427f5dc51d8d4a9332ef0/pkg/storage/replica.go#L1957)):*
> ```Go
> 			// queue so other waiters can proceed. This is safe because with
> 			// the stopper quiescing, no new commands can be proposed to
> 			// Raft successfully.
> ```

The reason this was necessary before was exactly the bug being fixed by this PR: `tryAbandon` was removing the proposal from `Replica.mu.proposals` which was preventing `endCmds.done` from being executed by `Replica.processRaftCommand`. Now that we're leaving the proposal in `Replica.mu.proposals` we can take the saner approach of always leaving invocation of `endCmds.done` to `Replica.processRaftCommand` as soon as we have proposed the command.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KYErRTDs-7pmom4sonX:b-toiggj)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
LGTM

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r8.
Review status: 2 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica.go, line 2175 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEvSkZaTJ3w0smoNp4:-KYEvSk_1sFadCASUcu5:b-nkoi8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a492779e4234a530c62cf1f7eea1a78338bbf112/pkg/storage/replica.go#L2175)):*
> ```Go
> 		cmd, ok := r.mu.proposals[idKey]
> 		if ok {
> 			cmd.Local.ctx = r.AnnotateCtx(context.Background())
> ```

as discussed: let's make this a TODO and ask @raduberinde about forking the span

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-:-KYEvXtCq-zLAlm4yLZo:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 2175 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000#-KYEvSkZaTJ3w0smoNp4:-KYEvu12ymbUElACzo8a:buepy9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a492779e4234a530c62cf1f7eea1a78338bbf112/pkg/storage/replica.go#L2175)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

as discussed: let's make this a TODO and ask @raduberinde about forking the span
</blockquote></details>

TODO added. @raduberinde do you want me to file an issue?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12000)*
<!-- Sent from Reviewable.io -->
